src/core/Solver_intf.ml

```diff
@@ -76,9 +76,19 @@ type ('term, 'formula, 'value) assumption =
 type ('term, 'formula, 'proof) reason =
   | Eval of 'term list
   (** The formula can be evalutaed using the terms in the list *)
-  | Consequence of 'formula list * 'proof
+  | Consequence of (unit -> 'formula list * 'proof)
   (** [Consequence (l, p)] means that the formulas in [l] imply the propagated
       formula [f]. The proof should be a proof of the clause "[l] implies [f]".
+
+      invariant: in [Consequence (l,p)], all elements of [l] must be true in
+      the current trail.
+
+      note on lazyiness: the justification is suspended (using [unit -> …])
+      to avoid potentially costly computations that might never be used
+      if this literal is backtracked without participating in a conflict.
+      However, if the theory isn't robust wrt backtracking an subsequent changes,
+      it can be easier to produce the explanation eagerly when
+      propagating, and then use [Consequence (fun () -> expl, proof)].
```

Comment:
Could you add a precision (and most definitely a corresponding comment in the `.ml` file) stating that since Consequences produce propagations at the current level, the suspended computation is only ever run in a "conservative extension" of the current theory state (because the propagation and associated suspended computation will be forgotten upon backtracking) and thus explains why the computation is relatively safe to suspend ?

Reply:
Note: "conservative extension" is the better word I could come up with to explain that the state is obtained after giving an arbitrary number of new slices, but not backtracking before the point where the propagation occurred. If you have a better explanation/wording, go for it, ^^

Reply:
Ok i'll consider these 2 points, indeed.